kubelet: ensure secret pulled image

pacoxu opened the pull request:
Rebased on #94899 by @mikebrow

/kind feature
/cc @mikebrow

```release-note
Implement support in kubelet to ensure images pulled with pod imagePullSecrets are authenticated
by other pods that do not have the same credentials.
```

##### Design Details: track a hash map for the credentials in kubelet

https://github.com/kubernetes/enhancements/pull/3532
> Kubelet will track, in memory, a hash map for the credentials that were successfully used to pull an image. It has been decided that the hash map will be persisted to disk, in alpha.

**What to store**
- image: string
- auth hash (keys): string
- ensuredBySecret: bool


**Where to store the data**
- store in node status (need to be confirmed with sig-node): **it needs API-change** to add a list field like `ensuredPulledImage` in node status spec.  
- store in a local file like `cpu policy state` (should it be secured by an encryption scheme). [No API change, but a new file on every node.]
- store in a secure volume? too complex?

**Should the stored data be outdated after several hours for security reasons?**
maybe 1, 6 hours, or 12 hours.

**Scenarios that should be taken care of **
1. kubelet restart (that is why we need to store the data)
2. pinned image: any special logic for a pinned image?

##### TODO List

* [x]  [ensure secret pulled images enhancements#1608](https://github.com/kubernetes/enhancements/pull/1608)
  code complete
  [x] [feature gate added for issue comment](https://github.com/kubernetes/enhancements/pull/1608#issuecomment-838805092)
* [ ]  usage (not needed for alpha)
* [ ]  docs (not needed for alpha)
* [x]  test bucket(s)
  unit buckets complete.. for FG on and off

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2535-ensure-secret-pulled-images/README.md
```

Test demo:
```
[root@daocloud ~]# bash -c "cat /var/lib/kubelet/image_manager_state | jq ."
{
  "images": {
    "sha256:eb6cbbefef909d52f4b2b29f8972bbb6d86fc9dba6528e65aad4f119ce469f7a": {
      "authHash": {
        "115b8808c3e7f073": {
          "ensured": true,
          "dueDate": "2023-05-26T18:00:09.802153792+08:00"
        }
      },
      "name": "daocloud.io/daocloud/dce-registry-tool:3.0.8"
    }
  }
}

[root@daocloud ~]# cat ensure.yaml
apiVersion: v1
kind: Pod
metadata:
  name: need-secret-but-has-no
spec:
  containers:
    - name: test-container
      image: daocloud.io/daocloud/dce-registry-tool:3.0.8

      command: [ "sleep", "100000" ]

---

apiVersion: v1
kind: Pod
metadata:
  name: need-secret-and-has-secret
spec:
  containers:
    - name: test-container
      image: daocloud.io/daocloud/dce-registry-tool:3.0.8
      command: [ "sleep", "100000" ]
  imagePullSecrets:
  - name: regcred

---

apiVersion: v1
kind: Pod
metadata:
  name: need-secret-and-has-wrong-secret
spec:
  containers:
    - name: test-container
      image: daocloud.io/daocloud/dce-registry-tool:3.0.8
      command: [ "sleep", "100000" ]
  imagePullSecrets:
  - name: regcred.1

[root@daocloud ~]#  kubectl get pod -o wide
NAME                               READY   STATUS             RESTARTS   AGE     IP               NODE       NOMINATED NODE   READINESS GATES
need-secret-and-has-secret         1/1     Running            0          2m5s    172.32.230.235   daocloud   <none>           <none>
need-secret-and-has-secret-2       1/1     Running            0          2m1s    172.32.230.234   daocloud   <none>           <none>
need-secret-and-has-wrong-secret   0/1     ImagePullBackOff   0          105s    172.32.230.236   daocloud   <none>           <none>
need-secret-but-has-no             0/1     ErrImagePull       0          70s     172.32.230.237   daocloud   <none>           <none>
```

mikebrow committed:
ensure secret pulled images; add PullImageSecretRecheckPeriod to option in

Signed-off-by: Mike Brown <brownwm@us.ibm.com>
Signed-off-by: Paco Xu <paco.xu@daocloud.io>

pacoxu commented:
/priority important-soon
/triage accepted

pacoxu commented:
Tested on a local env, and it can work. The dueDate for the ensured auth is now only 1 hour that can cover the kubelet restarts.

```
[root@daocloud ~]#  kubectl get pod -o wide
NAME                               READY   STATUS             RESTARTS   AGE     IP               NODE       NOMINATED NODE   READINESS GATES
need-secret-and-has-secret         1/1     Running            0          2m5s    172.32.230.235   daocloud   <none>           <none>
need-secret-and-has-secret-2       1/1     Running            0          2m1s    172.32.230.234   daocloud   <none>           <none>
need-secret-and-has-wrong-secret   0/1     ImagePullBackOff   0          105s    172.32.230.236   daocloud   <none>           <none>
need-secret-but-has-no             0/1     ErrImagePull       0          70s     172.32.230.237   daocloud   <none>           <none>
```

pacoxu commented:
/cc @bart0sh @SergeyKanzhelev

mikebrow reviewed (commented):
Looking good.. I think we need to remove entries with due date entries that have expired.

ruiwen-zhao commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -219,3 +322,57 @@ func applyDefaultImageTag(image string) (string, error) {
 	return image, nil
 }
+
+// isEnsuredBySecret - returns true if the secret for an auth used to pull an
```
can we update the comment here since it returns two boolean values?

pacoxu replied:
Updated.

ruiwen-zhao commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -167,6 +252,24 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
 	m.backOff.GC()
+
+	if m.state != nil {
+		m.lock.Lock()
```
if mutations to m.lock are guarded by mutex already, do we still need `m.lock.Lock()`?

pacoxu replied:
Removed.

ruiwen-zhao commented on pkg/kubelet/images/state_image.go:
```diff
@@ -0,0 +1,133 @@
+	if !ok {
+		digest.HashMatch[authHash] = &EnsuredBySecret{
+			Ensured: ensured,
+			DueDate: time.Now().Add(time.Hour),
```
should we be using the constant `ensuredExpirationDuration`, maybe define it here?

pacoxu replied:
Forget to use the constant that I defined.

ruiwen-zhao commented on pkg/kubelet/images/state_image.go:
```diff
@@ -0,0 +1,133 @@
+		}
+	} else {
+		ensuredBySecret.Ensured = ensured
+		ensuredBySecret.DueDate = time.Now().Add(time.Hour)
```
Do we ever evict a `digest.HashMatch[authHash]` entry when it passes DueDate? 

For example, if a secret-required image is being pulled more many times, each with a different AuthConfig (this could be due to credential rotation), will the hashmap keep expanding?

pacoxu replied:
Added the logic to do cleanup every 1 minute.

pacoxu commented on pkg/kubelet/cri/remote/remote_runtime_test.go:
```diff
@@ -110,6 +110,8 @@ func TestVersion(t *testing.T) {
 	rtSvc := createRemoteRuntimeService(endpoint, t)
 	version, err := rtSvc.Version(ctx, apitest.FakeVersion)
 	require.NoError(t, err)
-	assert.Equal(t, apitest.FakeVersion, version.Version)
```
Added from Mike's PR and I think this is not related. So I removed it now.

mikebrow commented on pkg/kubelet/images/state_checkpoint.go:
```diff
@@ -0,0 +1,182 @@
+	}
+}
+
+// ClearExpiredState clears state and saves it in a checkpoint
```
```suggestion
// ClearState clears state and saves it in a checkpoint
```

pacoxu replied:
Fixed.

mikebrow commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -63,30 +94,73 @@ func NewImageManager(recorder record.EventRecorder, imageService kubecontainer.I
+
+func (m *imageManager) StartGC() {
+	klog.InfoS("Starting Kubelet image pull secret ensuring GC manager")
+	go wait.Until(func() { m.state.ClearExpiredState() }, 1*time.Minute, wait.NeverStop)
```
seems aggressive to run this once a min... even if empty, even if the next timeout is closer to ensuredExpirationDuration = 1 * time.Hour 

Maybe if clear expired state returned the duration from now to the next timeout...

ruiwen-zhao replied:
+1 on making the GC interval 1 hour.

mikebrow reviewed (commented):
looking good!

suggest returning a boolean if clearexpiredstate actually cleared an item and only storeState state if so..

pacoxu commented:
/test pull-kubernetes-e2e-gce-cos-alpha-features

pacoxu commented:
ACK, code freeze in three weeks.

Can I get another round of review?

haircommander commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -63,30 +94,73 @@ func NewImageManager(recorder record.EventRecorder, imageService kubecontainer.I
+			(container.ImagePullPolicy == v1.PullIfNotPresent && (!imagePresent)) {
+			return true
+		}
+	} else { // new gated behavior
```
nit: I'm personally more of a fan of more return statements rather than having a bunch of nested elses. up to you

mikebrow replied:
I'd prefer to use a switch on the gate.. was just trying to make it easier for reviewers to read the policy change

pacoxu replied:
refactor the func code in https://github.com/kubernetes/kubernetes/pull/114847/commits/28a55af28703b0d8c465c619673456b5bcba0a95 to make it more clear.

Use `switch` for `ImagePullPolicy` and return early in the below logic.

haircommander commented on pkg/kubelet/container/helpers.go:
```diff
@@ -134,6 +135,14 @@ func HashContainerWithoutResources(container *v1.Container) uint64 {
+// HashAuth - returns a hash code for a CRI pull image auth
+func HashAuth(auth *runtimeapi.AuthConfig) string {
+	hash := fnv.New64a()
+	authJSON, _ := json.Marshal(auth)
```
can you add a comment as to why we ignore the error here?

pacoxu replied:
Updated. log an error and return "" here.

haircommander commented on pkg/kubelet/images/checkpoint.go:
```diff
@@ -0,0 +1,60 @@
+
+// VerifyChecksum is place holder to do verify
+func (cp *ImageManagerCheckpointV1) VerifyChecksum() error {
+	return nil
```
do we have a note on when we're dealing with this place holder?

pacoxu replied:
Add the logic.

haircommander commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -219,3 +317,58 @@ func applyDefaultImageTag(image string) (string, error) {
+// be removed from the ensured map once it is expired.
+// 2. ensuredBySecret returns true if the secret for an auth used to pull an
+// image has already been authenticated through a successful pull request
+// and the same auth exists for this podSandbox/image/.
```
I'm not sure you mean to keep the second `/` in `podSandbox/image/`

pacoxu replied:
Updated.

haircommander commented on pkg/kubelet/images/image_manager_test.go:
```diff
@@ -631,14 +631,18 @@ func TestShouldPullImage(t *testing.T) {
-		if utilfeature.DefaultFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) {
-			assert.Equal(t, test.expectedWithFGOn, sp, "TestCase[%d]: %s ensured image fg enabled", i, test.description)
-		} else {
+	t.Run("disabled_features", func(t *testing.T) {
```
nit: is there a reason not to mention which feature? I only see one being disabled/enabled

pacoxu replied:
Add the feature gate name here.

haircommander commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -173,7 +204,11 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
+	var pulledBySecret, ensuredBySecret bool
+	if present {
+		pulledBySecret, ensuredBySecret = m.isEnsuredBySecret(imageRef, spec, pullSecrets)
+		klog.V(5).InfoS("Get ensured check by securet", "image", image, "imageRef", imageRef, "pulledBySecret", pulledBySecret, "ensuredBySecret", ensuredBySecret)
```
spellcheck: secret*

haircommander commented on pkg/kubelet/images/state.go:
```diff
@@ -0,0 +1,55 @@
+limitations under the License.
+*/
+
+package images
```
personally, I think this file could be named something clearer--it's holding very specific state

pacoxu replied:
renamed to `images/ensure_pull_stats.go`. Not sure if it is clearer.

pacoxu commented:
Sorry for late update due to missing comments notificaiton.

haircommander commented:
/lgtm

pacoxu commented:
/assign @mikebrow @mrunalp @dchen1107

pohly commented on pkg/kubelet/images/image_manager_test.go:
```diff
@@ -241,17 +244,22 @@ func pullerTestEnv(t *testing.T, c pullerTestCase, serialized bool, maxParallelI
 	fakeRuntime.InspectErr = c.inspectErr
 
 	fakePodPullingTimeRecorder = &mockPodPullingTimeRecorder{}
+	tmpStateDir, err := os.MkdirTemp("", "image_manager_state")
```
Would it be possible to use `testing.T.TempDir`? Then you don't need error checking and the directory will be removed automatically for you.

pacoxu replied:
Good catch. Updated.

haircommander commented:
I know we risk bike shedding here, but I had some thoughts about this:
- I don't understand the motivation to do the checkpointing. If the image is already present, but we have PullPolicy: Always, then the runtime *should* (based on my understanding) just auth with the server, and not actually pull any blobs redundantly. What's the risk of redoing that work over kubelet reboots?
- If we do keep the checkpoint file: I can't see quickly where the location is. Are we keeping it on disk? if so, can we put it somewhere in a tmpfs instead? I don't think we need it across node reboots

Happy to discuss on a call or something if that'd be better, sorry if I'm missing some context

mikebrow commented:
> I know we risk bike shedding here, but I had some thoughts about this:
> 
> * I don't understand the motivation to do the checkpointing. 

Motivation Discussion here:
https://github.com/kubernetes/kubernetes/pull/94899#issuecomment-1058443476

> If the image is already present, but we have PullPolicy: Always, then the runtime _should_ (based on my understanding) just auth with the server, and not actually pull any blobs redundantly.

correct, if every pod is using pull always or pull never policy or if the tags are `:latest` then no change

The performance and availability problem with that is having to connect and check manifests every time for every container image.  Additionally, at least one container registry charges(counts) the connect and manifest check as one of your container image pulls, even if you don't pull blobs. Further, the policy pull if not present policy is still an option and is not doing what users expect, and there is a desire to have a means to discover if an image was preloaded for disconnected use / performance reasons or pulled with pull always or pull if not present.

> What's the risk of redoing that work over kubelet reboots?

which images are pre-loaded images, which were pulled with a particular password and thus are considered only accessible with that access right...

> * If we do keep the checkpoint file: I can't see quickly where the location is. Are we keeping it on disk? if so, can we put it somewhere in a tmpfs instead? 

@pacoxu ?

> I don't think we need it across node reboots

If we reset (gc) the image cache maybe.. but that was rejected ^

> Happy to discuss on a call or something if that'd be better, sorry if I'm missing some context

Sure let's chat!

haircommander commented:
> disconnected use / performance reasons

I don't see how disconnected works with this TBH so I am unsure why it's a case to consider. Perf reasons is a reason to keep the checkpoint file, but enabling this feature generally is trading off perf and security. I guess the question is: is the added code of keeping the checkpoint worth saving for the (IMO ideally infrequent) situation of a kubelet restart?

> If we reset (gc) the image cache maybe.. but that was rejected ^

if the checkpoint file isn't present (if put in a tmpfs) then kubelet would treat every auth request as invalidated and rebuild the checkpointing. I don't think we need to clean the image cache to get this behavior.

haircommander commented:
>  I also believe we should tackle issues related to registry connection or registry failure time (the registry may crash as well). Additionally, the kubelet may be restarted as the same time.

I think this needs to be a calculated decision made by an administrator. This feature will reduce uptime in some scenerios and they should be aware of that/have registry mirrors configured to account for that. I think there should be a kubelet config field to toggle this behavior so that even after the feature goes GA an admin can opt-out

pacoxu commented:
I just watched the record.

To summarize:
1. Current logic(without this PR)
User has to use `Always` as current IfNotPresent will never to re-check.
No Doubt 

2. Wanted Logic(Secure way using IfNotPresent):

Skip re-check for checked secret in a period of 1h if kubelet is not restart.

3. Add a new flag/kubelet configuration param (Derek inspired):
A optional `PullImageSecretRecheckDuration` to configure the recheck duration
- 0 for always re-check: most secure one.
- -1: for never re-check: Set this to default to make no behavior change. But not secure at all.
- a duration like 1h, 1d, 1m, 1week: after setting it, administrators can use IfNotPresent instead of Always then and the administors knows the trade-off.

Then administrator can know that he have a cache buffer not to re-check secret.

4. The Debate(use checkpoint or not): Should we skip the re-check after kubelet restart or reboot?
 
I think that we may not use checkpoint. Reasons:
 - Restart Scenarios: 
   - upgrade: user needs to drain the node according to the best practice, and re-check is acceptable. (Honestly, many users don't)
  - other problems(change a configuration or some scripts for memory leak): still some maintenance modes, and re-check is acceptable.
- When no-checkpoints hurts?
  - Continuous kubelet restarts（this is a bug scanerio）
 
 
---


- Should we make the PullImageSecretRecheckDuration configurable? I think yes.
- Which should be the default behavior of IfNotPresent in the future? Always re-check? Never re-check? Re-check with 1h expired?  I prefer always recheck as it is most secure one but a behavior change.
- Should we add a flag to support the checkpoint here for restarts/reboots? Just remove the checkpoint part? I prefer removing checkpoints. 
 
@mikebrow @haircommander correct me if I understand wrongly and give your opinions.

haircommander commented:
Great summary. One point to clarify:  I think the IfNotPresent behavior will not necessarily always recheck. I think kubelet should still keep its internal hash map, and if the request uses the same creds again (pod is restarting or same user is creating multiple replicas) then the image shouldn't repull, until the duration expires. My point to drop the checkpoint was only to get rid of saving it in the file, for the reasons you outlined above. Thoughts??

pacoxu commented:
> I think kubelet should still keep its internal hash map, and if the request uses the same creds again (pod is restarting or same user is creating multiple replicas) then the image shouldn't repull, until the duration expires.

So we should have a new flag like `PullImageSecretRecheckDuration` to make the expired duration configurable. 
For `IfNotPresent`image, we check the internal hash map and it can also expire. 

I agree with the part definitly.

---

For the restart scenario, we will drop the checkpoint and do the recheck. 
- we add persist part in https://github.com/kubernetes/enhancements/pull/3532  @mrunalp @mikebrow 

@derekwaynecarr @dchen1107  Do you agree on removing the checkpoint part?

@mikebrow I prefer to not use persist in the alpha stage and get more feedback later to decide if we will add the persist for this feature.

pacoxu commented:
I updated kep in https://github.com/kubernetes/enhancements/pull/4431 and I will update this PR later.

pacoxu commented:
I updated the PR
- drop the checkpoint part
- add new kubelet configuration `pullImageSecretRecheckDuration`

/cc @mikebrow @haircommander 
do you have time to take a look again?

kannon92 commented:
@pacoxu can you update this PR to fix CI failures?

pacoxu commented:
> @pacoxu can you update this PR to fix CI failures?

Fixed.

kannon92 commented:
> > @pacoxu can you update this PR to fix CI failures?
> 
> Fixed.

Still have a failure in pull-kubernetes-linter-hints. Looks like you used pointer rather than ptr in some places.

rphillips commented on pkg/kubelet/apis/config/fuzzer/fuzzer.go:
```diff
@@ -119,6 +119,7 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
 			obj.EnableSystemLogHandler = true
 			obj.MemoryThrottlingFactor = utilpointer.Float64(rand.Float64())
 			obj.LocalStorageCapacityIsolation = true
+			obj.PullImageSecretRecheckDuration = metav1.Duration{Duration: 24 * time.Hour}
```
This 24h duration is used in a few spots. Can you create a constant for the RecheckDuration?

pacoxu replied:
Add a `DefaultRecheckDuration`.

haircommander commented on pkg/kubelet/images/image_manager_test.go:
```diff
@@ -241,17 +244,17 @@ func pullerTestEnv(t *testing.T, c pullerTestCase, serialized bool, maxParallelI
 }
 
 func TestParallelPuller(t *testing.T) {
+	defer os.RemoveAll(tmpStateDir)
```
it's not clear to me why we're removing this in a test that doesn't apparently use it

pacoxu replied:
This needs to be removed. It used for checkpoint saving before.

rphillips commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -224,3 +308,68 @@ func applyDefaultImageTag(image string) (string, error) {
+// image has already been authenticated through a successful pull request
+// and the same auth exists for this podSandbox/image.
+func (m *imageManager) isEnsuredBySecret(imageRef string, image kubecontainer.ImageSpec, pullSecrets []v1.Secret) (pulledBySecret, ensuredBySecret bool) {
+	m.lock.Lock()
```
This function ignores all the errors. We should either log the error or (probably better) return the error so the caller can log it.

pacoxu replied:
Add an err return, and log the err.

harche replied:
@pacoxu I see that comment has been resolved, but the code in that function still neither logs nor returns the error.

haircommander commented:
besides the nodes and the linter issue this LGTM

harche commented on pkg/kubelet/container/helpers.go:
```diff
@@ -134,6 +135,18 @@ func HashContainerWithoutResources(container *v1.Container) uint64 {
+	authJSON, err := json.Marshal(auth)
+	if err != nil {
+		klog.ErrorS(err, "Failed to marshal auth config")
+		return ""
```
Why to suppress the error and return empty string? Why not return the (string, error)?

I see that this has been used in https://github.com/kubernetes/kubernetes/pull/114847/files#diff-b22ab169cdc70b2e968bb653f29cf0a1384d3bccaebeef96aebf2a5a40821bbeR367. But there is an already comment by @rphillips on that part of the code to return the error.

pacoxu replied:
Fixed.

harche commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -136,12 +185,32 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
+	if present {
+		pulledBySecret, ensuredBySecret, err := m.isEnsuredBySecret(imageRef, spec, pullSecrets)
+		if err != nil {
+			klog.V(4).ErrorS(err, "Error get ensured check by secret", "image", image, "imageRef", imageRef, "pulledBySecret", pulledBySecret, "ensuredBySecret", ensuredBySecret)
```
Should we return the error to the caller instead of just logging?

pacoxu replied:
Yes.

harche commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -136,12 +185,32 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
+				// TODO: add integration test for this thrown error message
+				msg := fmt.Sprintf("Container image %q is present with pull policy of %q but does not have the proper auth (image secret) that was used to pull the image", container.Image, container.ImagePullPolicy)
+				m.logIt(ref, v1.EventTypeWarning, events.ErrImageNotEnsured, logPrefix, msg, klog.Warning)
+				return "", msg, ErrImageNotEnsured
```
The function returns (imageRef, error message, error), you already have the `ref`, why not to use it while returning instead of ""  ?

pacoxu replied:
Yeah, we can return the `imageRef` here and below.

harche commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -70,16 +70,20 @@ func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecon
-			return imageRef, nil
+			// also return the hash for the auth for this successful pull of this imageRef
+			// hash may be used to ensure another use of the imageRef is also authorized
+			hash := kubecontainer.HashAuth(auth)
```
`HashAuth()` can return empty string if it encounters the error, so should we check for the condition here? Also what do you think of `HashAuth()` returning (string, error) so that we can propagate the error to the caller instead of just logging it?

pacoxu replied:
changed to `returning (string, error) `.

harche commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -224,3 +312,68 @@ func applyDefaultImageTag(image string) (string, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if imageRef == "" {
+		return
```
As mentioned by @rphillips earlier, should ideally return the error or at least log it.

pacoxu replied:
Fixed.

pacoxu commented:
@harche  Thanks for your review.
Latest comments are addressed in https://github.com/kubernetes/kubernetes/compare/5bc731fc6d5a9701b129c5effc3ab26e4c1eb45d..14570664618938bd33eabbcdb2052da8ddba5256.

pacoxu commented:
@haircommander  @harche @mikebrow @rphillips 
Do you have time to take a look again before code freeze?

haircommander commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -18,24 +18,49 @@ package images
+	lastEnsuredDate time.Time
+}
+
+type ensureSecretPulledImageDigest struct {
```
nit: it's my preference to define structures/functions below where they're used so files can be read top-down. Can this be switched with ensuredInfo?

pacoxu replied:
Moved.

haircommander commented:
a nit and you have linter errors.

One more note: I think technically `PullImageSecretRecheckPeriod` would be more correct, as duration implies the recheck is happening for the whole duration, rather than periodically. I know it's late and I am not stubborn about this, but if you find time before this merges I think I'd prefer it everywhere *Duration is used as a name here

pacoxu commented:
> One more note: I think technically `PullImageSecretRecheckPeriod` would be more correct, as duration implies the recheck is happening for the whole duration, rather than periodically. I know it's late and I am not stubborn about this, but if you find time before this merges I think I'd prefer it everywhere *Duration is used as a name here

Good catch, we have `VolumeStatsAggPeriod` and `CPUManagerReconcilePeriod` and so on. 

PullImageSecretRecheckDuration -> PullImageSecretRecheckPeriod
LGTM

pacoxu commented:
> a nit and you have linter errors.

Fixed.
 
> One more note: I think technically `PullImageSecretRecheckPeriod` would be more correct, as duration implies the recheck is happening for the whole duration, rather than periodically. I know it's late and I am not stubborn about this, but if you find time before this merges I think I'd prefer it everywhere *Duration is used as a name here

`PullImageSecretRecheckDuration` -> `PullImageSecretRecheckPeriod`

haircommander commented on pkg/kubelet/apis/config/v1beta1/defaults.go:
```diff
@@ -39,6 +39,8 @@ const (
 	// See https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos
 	DefaultMemoryThrottlingFactor = 0.9
+
+	DefaultRecheckDuration = 24 * time.Hour
```
a left over `DefaultRecheckDuration`. also can this be a bit more clear like 
`DefaultImagePullRecheckPeriod`? More verbose but since it's exported and used elsewhere I think that'll make it clearer

pacoxu replied:
Updated and add an link to the KEP here like above.

mrunalp commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -164,14 +233,33 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
+		digest := m.ensureSecretPulledImage[imageRef]
+		if digest == nil {
+			digest = &ensureSecretPulledImageDigest{HashMatch: make(map[string]*ensuredInfo)}
+			m.ensureSecretPulledImage[imageRef] = digest
```
Shouldn't the key of this map be a list of auth digests? A single image could have multiple auths for different users. cc: @mikebrow

pacoxu replied:
`ensureSecretPulledImage` is a map `map[string]*ensureSecretPulledImageDigest`.
- The map key is image ref.
- The key is `ensureSecretPulledImageDigest` which contains hashMap `HashMatch map[string]*ensuredInfo`.
  - the key is auth's hash
  - the value is `&ensuredInfo{ensured=true, lastEnsuredDate=time.Now()} `

So a single image can have multi auths here.

mikebrow replied:
yeah probably will be easier to read with the proposed name changes.. :-)

haircommander commented:
/lgtm

mrunalp commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -64,27 +79,44 @@ func NewImageManager(recorder record.EventRecorder, imageService kubecontainer.I
+	case v1.PullAlways:
 		return true
+	case v1.PullIfNotPresent:
+		if !utilfeature.DefaultFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) {
```
nit: This may read easier: 
```
	case v1.PullIfNotPresent:
		// pull if image doesn't exist
		if !imagePresent {
			return true
		}
		// if the imageRef has been pulled by a secret and Pull Policy is PullIfNotPresent
		// we need to ensure that the current pod's secrets map to an auth that has Already
		// pulled the image successfully. Otherwise pod B could use pod A's images
		// without auth. So in this case if pulledBySecret but not ensured by matching
		// secret auth for a pull again for the pod B scenario where the auth does not match
		if utilfeature.DefaultFeatureGate.Enabled((features.KubeletEnsureSecretPulledImages)) {
			if pulledBySecret && !ensuredBySecret {
				return true // noting here that old behaviour returns false in this case indicating the image should not be pulled
			}
		}
		return false
```

mikebrow replied:
agreed that reads cleaner .. because the feature gate only applies to present images..

mrunalp commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -97,6 +129,24 @@ func (m *imageManager) logIt(ref *v1.ObjectReference, eventtype, event, prefix,
+	//       note @mrunalp makes a good point that we can utilize /apimachinery/pkg/util/sets/string.go here
+
+	// map of auth hash (keys) used to successfully pull this imageref
+	HashMatch map[string]*ensuredInfo
```
nit: `Auths`

mrunalp commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -64,27 +79,44 @@ func NewImageManager(recorder record.EventRecorder, imageService kubecontainer.I
+		puller:                       puller,
+		podPullingTimeRecorder:       podPullingTimeRecorder,
+		keyring:                      keyring,
+		ensureSecretPulledImage:      make(map[string]*ensureSecretPulledImageDigest),
```
nit: `ensureSecretPulledImages`

mrunalp commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -97,6 +129,24 @@ func (m *imageManager) logIt(ref *v1.ObjectReference, eventtype, event, prefix,
+	lastEnsuredDate time.Time
+}
+
+type ensureSecretPulledImageDigest struct {
```
nit: Maybe `imagePullInfo`

pacoxu commented:
Fixed some naming and code reading problem. @mrunalp 
Not sure if I miss the code freeze😄

kannon92 commented:
> Fixed some naming and code reading problem. @mrunalp Not sure if I miss the code freeze😄

Yea, I think it was at 1800 PST (38 minutes ago).

MaryamTavakkoli commented:
/milestone v1.30

dchen1107 commented:
/milestone v1.30

dchen1107 commented:
/lgtm

dchen1107 commented:
/approve

kannon92 commented:
/label api-review

liggitt commented on staging/src/k8s.io/kubelet/config/v1beta1/types.go:
```diff
@@ -851,6 +851,10 @@ type KubeletConfiguration struct {
+
+	// PullImageSecretRecheckPeriod defines the duration to recheck the pull image secret.
+	// By default, the kubelet will recheck the pull image secret every 24 hours(1d).
+	PullImageSecretRecheckPeriod *metav1.Duration `json:"pullImageSecretRecheckPeriod"`
```
Is this the only config surface? I'm surprised we're expecting to default this tightening on when the feature graduates to beta. I expected the other way around, default to 0 / off, but let admins opt in.

mrunalp replied:
sounds good! However bad the current behavior may be, we don't to break pods inadvertantly relying on it so opt in makes sense.

mikebrow replied:
I concur with a note:

The reasoning was as follows:

Currently admins desiring to be secure by default in a multi user or reduction of permission to access an image scenario, were forced into restricting away the if not present policy by employing the AlwaysPullImages admission controller https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#alwayspullimages ...
The admins could also require `:latest tags`.. or employ a scheme to gc images.

Thus when we make the default off we will be asking admins to "trust" IfNotPresent now processing only if you they explicitly enable secure IfNotPresent behavior.

pacoxu replied:
Yes, this will be a behavior change when it promoted to beta if we default it to 24h. I think we wanted to make it a secure way when we discussed about the default behavior here. I agree we might not break current users who is using `IfNotPresent` and this new behavior may caused that the pod crash after this recheck.

So we will change the default value to 0. It means that we don't recheck the secret again for `IfNotPresent` ImagePullPolicy.
```
	// if the pullImageSecretRecheckPeriod is 0, we don't need to check the secret again
	if m.pullImageSecretRecheckPeriod.Duration == 0 {
		return pulledBySecret, true, nil
	}
```

I will update the PR.

pacoxu replied:
> Thus when we make the default off we will be asking admins to "trust" IfNotPresent processing only if they explicitly enable secure IfNotPresent behavior.

BTW, can we add a note that suggests user to use a recheck period to be more secure?

liggitt replied:
default change and comment lgtm. traced through to where this is used and had a question about whether the *re*validation duration is what we want to decide whether to *ever* validate a pull secret

pacoxu committed:
set PullImageSecretRecheckPeriod default to 0; default not recheck for IfNotPresent Image

pacoxu commented:
I changed the Default value in  https://github.com/kubernetes/kubernetes/compare/561eb58389377daf3d51d23ccf67e3792366a2eb..5acdb0defcb9cc2cf9318ced3318e562f48358b9.

```
	// PullImageSecretRecheckPeriod defines the duration to recheck the pull image secret.
	// By default, the kubelet will not recheck the pull image secret.
	// For security reasons, we recommend rechecking the pull image secret, ideally every 24 hours (1d).
```
Adding a comment like this.

salehsedghpour commented:
As the exception request is approved, I'm adding a `/hold` label to prevent from accidental merge.
/hold

pacoxu commented:
/test pull-kubernetes-e2e-gce

liggitt commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -70,16 +70,23 @@ func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecon
 		// If there was no error, return success
 		if err == nil {
-			return imageRef, nil
+			// also return the hash for the auth for this successful pull of this imageRef
+			// hash may be used to ensure another use of the imageRef is also authorized
+			hash, err := kubecontainer.HashAuth(auth)
+			if err != nil {
+				klog.ErrorS(err, "Failed to hash auth", "image", img)
+			}
```
only run this code when the feature is enabled

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -32,16 +32,16 @@ import (
 	repoToPull, _, _, err := parsers.ParseImageName(img)
 	if err != nil {
-		return "", err
+		return "", "", err
```
```suggestion
                retErr = err
		return
```

mikebrow commented on pkg/kubelet/container/runtime.go:
```diff
@@ -149,8 +149,9 @@ type StreamingRuntime interface {
-	PullImage(ctx context.Context, image ImageSpec, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig) (string, error)
+	// secrets if necessary. It returns a reference (digest or ID) to the pulled image
+	// and if applicable a hash for the auth used to successfully pull the image.
+	PullImage(ctx context.Context, image ImageSpec, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig) (string, string, error)
```
clears up but requires named return style in func

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -32,16 +32,16 @@ import (
 	keyring, err := credentialprovidersecrets.MakeDockerKeyring(pullSecrets, m.keyring)
 	if err != nil {
-		return "", err
+		return "", "", err
```
```suggestion
                retErr = err
		return
```

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -50,13 +50,13 @@ func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecon
 		klog.V(3).InfoS("Pulling image without credentials", "image", img)
 
+		// callout to cri impl of pull image
 		imageRef, err := m.imageService.PullImage(ctx, imgSpec, nil, podSandboxConfig)
```
```suggestion
		imageRef, retErr = m.imageService.PullImage(ctx, imgSpec, nil, podSandboxConfig)
```

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -50,13 +50,13 @@ func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecon
 
+		// callout to cri impl of pull image
 		imageRef, err := m.imageService.PullImage(ctx, imgSpec, nil, podSandboxConfig)
 		if err != nil {
```
```suggestion
		if retErr != nil {
```

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -50,13 +50,13 @@ func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecon
 		if err != nil {
 			klog.ErrorS(err, "Failed to pull image", "image", img)
-			return "", err
+			return "", "", err
```
```suggestion
			return
```

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -50,13 +50,13 @@ func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecon
+		// callout to cri impl of pull image
 		imageRef, err := m.imageService.PullImage(ctx, imgSpec, nil, podSandboxConfig)
 		if err != nil {
 			klog.ErrorS(err, "Failed to pull image", "image", img)
```
```suggestion
			klog.ErrorS(retErr, "Failed to pull image", "image", img)
```

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -50,13 +50,13 @@ func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecon
 		}
-
-		return imageRef, nil
+		return imageRef, "", nil
```
```suggestion
		return
```

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -70,16 +70,23 @@ func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecon
 		}
 
+		// callout to cri impl of pull image with auth
 		imageRef, err := m.imageService.PullImage(ctx, imgSpec, auth, podSandboxConfig)
```
```suggestion
		imageRef, retErr = m.imageService.PullImage(ctx, imgSpec, auth, podSandboxConfig)
```

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -70,16 +70,23 @@ func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecon
+		// callout to cri impl of pull image with auth
 		imageRef, err := m.imageService.PullImage(ctx, imgSpec, auth, podSandboxConfig)
 		// If there was no error, return success
 		if err == nil {
```
```suggestion
		if retErr == nil {
```

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -70,16 +70,23 @@ func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecon
-			return imageRef, nil
+			// also return the hash for the auth for this successful pull of this imageRef
+			// hash may be used to ensure another use of the imageRef is also authorized
+			hash, err := kubecontainer.HashAuth(auth)
```
```suggestion
			hash, retErr = kubecontainer.HashAuth(auth)
```

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -70,16 +70,23 @@ func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecon
+			// also return the hash for the auth for this successful pull of this imageRef
+			// hash may be used to ensure another use of the imageRef is also authorized
+			hash, err := kubecontainer.HashAuth(auth)
+			if err != nil {
```
```suggestion
			if retErr != nil {
```

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -70,16 +70,23 @@ func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecon
+			// hash may be used to ensure another use of the imageRef is also authorized
+			hash, err := kubecontainer.HashAuth(auth)
+			if err != nil {
+				klog.ErrorS(err, "Failed to hash auth", "image", img)
```
```suggestion
				klog.ErrorS(retErr, "Failed to hash auth", "image", img)
```

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -70,16 +70,23 @@ func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecon
+			if err != nil {
+				klog.ErrorS(err, "Failed to hash auth", "image", img)
+			}
+			return imageRef, hash, nil
```
```suggestion
			return
```

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -70,16 +70,23 @@ func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecon
+			return imageRef, hash, nil
 		}
 
 		pullErrs = append(pullErrs, err)
```
```suggestion
		pullErrs = append(pullErrs, retErr)
```

mikebrow commented on pkg/kubelet/kuberuntime/kuberuntime_image.go:
```diff
@@ -70,16 +70,23 @@ func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecon
 	}
 
-	return "", utilerrors.NewAggregate(pullErrs)
+	return "", "", utilerrors.NewAggregate(pullErrs)
```
```suggestion
        imageRef = ""
        hash = ""
        retErr = utilerrors.NewAggregate(pullErrs)
	return
```

mikebrow reviewed (commented):
see proposed style change to named return vals...

@pacoxu pls ignore the this style change.. diff is to big.. see comment below from Jordan

liggitt commented:
> see proposed style change to named return vals...

I'd rather minimize the diff... returning data (imageRef) alongside errors makes me wonder what is using that data in error cases

mikebrow commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -224,3 +312,68 @@ func applyDefaultImageTag(image string) (string, error) {
+	}
+
+	// if the image is in the ensured secret pulled image list, it is pulled by secret
+	pulledBySecret := m.ensureSecretPulledImages[imageRef] != nil
```
checkpointing was out then in .. now out again for the alpha.. TODO: restore from checkpoint..

liggitt replied:
> checkpointing was out then in .. now out again for the alpha.. TODO: restore from checkpoint..

Ok, sorry for asking questions which have already been answered :-)

Seems like checkpointing is the hard part of the problem. I'll defer to node approvers on that decision, but I'm a little surprised we're deferring tackling it.

haircommander replied:
I was the one that proposed we drop the checkpointing. The motivation was the recheck period was proposed to be 24hr, and I think the kubelet restarting is an infrequent operation. Redundantly checking the creds is part of the feature's intention, so the admin is opting into checking more than necessary and that's the point. Not having checkpointing reduces the code complexity while not materially changing the behavior IMO

liggitt replied:
If I'm reading correctly, the current implementation means images previously pulled to the node using pull secrets are treated as non-secret-pulled images after kubelet restart and would never be required to revalidate their pull credentials. Am I reading that correctly?

pacoxu replied:
I had a summary about the behavior in https://github.com/kubernetes/kubernetes/pull/114847#issuecomment-1857559176.
- Current Behavior: User has to use Always as current IfNotPresent will never do re-check.

haircommander replied:
I think the problem @liggitt is describing is the ensureSecretPulledImages struct is only populated at the end of an EnsureImageExists call, which is after the first time it'd be used after the kubelet restarts. If that map is empty (because we have yet to try to pull an image) wouldn't the kubelet think it never needed to pull the image with a secret? an empty map is basically treated as no images required secret pulls, so the first time an image is requested after the recheck period or after kubelet restart would get no auth

mikebrow replied:
> wouldn't the kubelet think it never needed to pull the image with a secret?

if the gate is off .. ifNotPresent means, in the context of ensuring images for a new pod, don't recheck auth/manifests unless the tag is :latest  thus an admin must force pull always to check/recheck auth 

if the gate is on .. ifNotPresent means, in the context of ensuring images for a new pod, check/recheck within check option configuration parameters before starting a new pod as a part of the ensureImage process.. IOW a restart without checkpoint merely means we have to be more aggressive on the first pod ensureImage checks and go ahead and do the check to initialize the ensured/hash cache..

mrunalp replied:
Falling back to recheck on empty map makes sense. However, we should document that it could break the use case of someone installing a k8s cluster with all image IfNotPresent and then using it disconnected from the registry. The user would see issues on kubelet/node restarts because the rechecks will start failing. This case will need checkpointing to be implemented.

liggitt replied:
> Falling back to recheck on empty map makes sense. However, we should document that it could break the use case of someone installing a k8s cluster with all images `IfNotPresent` and then using it disconnected from the registry. The user would see issues on kubelet/node restarts because the rechecks will start failing.

Exactly. This "how do we know which pre-existing images are intentionally preloaded and which are previously-pulled-by-secret" question is the hard part of the problem. That's why I was surprised to see this first version not handling it at all.

haircommander replied:
> Falling back to recheck on empty map makes sense. However, we should document that it could break the use case of someone installing a k8s cluster with all images IfNotPresent and then using it disconnected from the registry. The user would see issues on kubelet/node restarts because the rechecks will start failing. This case will need checkpointing to be implemented.

this is true, but also if this is the case then that admin probably shouldn't enable this feature, because in addition to not handling kubelet restarts, there's also a recheck period. I suppose the recheck period could be arbitrarily high (never recheck) but it would be odd if a security minded admin wanted this feature but didn't utilize a lower frequency...

in other words: kubelet checkpointing could prevent some rechecks, but the point of this KEP is to add unnecessary rechecks. An admin should only enable it if their registries are expected to be available

mikebrow replied:
https://github.com/kubernetes/kubernetes/pull/94899/files#r691688013 prior discussion on this context

mikebrow replied:
> the point of this KEP is to add unnecessary rechecks

original point was to provide a path to get off pull always

The desire to do intermittent rechecks (e.g. after kubelet restart or after duration) for better security than if not present and better performance over pull always is at conflict with the desire to not have any rechecks because a recheck might fail in cases where a user does not use pull always or pull never and is expecting pull if not present to behave like pull never. 

With rechecking enabled for any reason in the scope of the if not present policy the checkpointing has minimal value.

With rechecking disabled for any reason in the scope of the if not present policy.. the checkpointing is needed before we go beta otherwise "how do we know which pre-existing images are intentionally preloaded and which are previously-pulled-by-secret."

So it would seem... the resolution for alpha would be to check if either of the recheck signals are enabled and if not .. disable the feature gate and possibly warn users that they should employ the pull always admission controller.

pacoxu replied:
https://github.com/kubernetes/kubernetes/pull/114847/commits/7a120ec8850900a419db335db37cf7063cba3b3f

if pullImageSecretRecheckPeriod>0 and KubeletEnsureSecretPulledImages is enabled, do recheck related logic.

This will keep the original behavior if we don't set the recheck period even when we enabled the FG by default in later releases or manually enable it in v1.30.

Does this make sense?

liggitt replied:
I'm actually out of the office today, so I can't give this more time until next week, the more I look at this, I don't see how this can work as expected (reliably make new pods/pulls prove they have access AND not destabilize already running/working pods if a registry goes unavailable) without kubelet keeping track across restarts of images it used pull credentials to fetch.

There's two questions the kubelet has to answer:
1. did image X which is already pulled to the node require image pull credentials?
2. if so, are the image pull credentials this pod requesting image X is providing good?

This PR assumes all images that exist at kubelet start don't require credentials, only keeps track of credential-enabled image pulls the kubelet does in-memory, then forgets them on restart.

The "periodically check" approach looks like it will not actually protect images that exist on startup from use by uncredentialed pods. That seems like a fundamental gap.

The "periodically check" approach also looks like it will put existing working credentialed pods at risk of registry outage. Is it actually intended that pods which had image pull credentials and proved their pull credential was valid on their initial start on a node could stop working later because the registry is unavailable? That doesn't seem right.

https://github.com/kubernetes/kubernetes/pull/94899#discussion_r818144101 and https://github.com/kubernetes/kubernetes/pull/94899#issuecomment-1058126793 were the feedback on the last attempt of this feature, and looks like we're in essentially the same position now.

pacoxu replied:
The two stories from the kep https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2535-ensure-secret-pulled-images#user-stories.
1. Story 1: User with multiple tenants will be able to support all image pull policies without concern that one tenant will gain access to an image that they don't have rights to.
2.Story 2: User will will no longer have to inject the Pull Always Image Pull Policy to ensure all tenants have rights to the images that are already present on a host.

The aim of the KEP is to gain an image pull policy that is between `Always` and current `IfNotPresent`.

- Image pull policy`Always` trust nothing.
- Image pull policy`IfNotPresent` trust everything on the disk, but no checking on the image pull secret.
- What we want is something like`IfNotPresentWithTrustedAuthPeriod`: 
  - we add a `recheck/trust period` to control the weak level: 24h is like a daily trust credit. 1y may be a long credit trust. Node admin can control its trust for an image pull credential check result.


The `PullImageSecretRecheckPeriod` naming is somewhat inaccurate, as we didn't recheck the image secret every 12h if it is set to be 12h. The check will happen only when a container startup or restart happen IIUC. It will not kill a pod or container if it found the credential is out-of-dated.

What should be done?
1. We should keep the behavior not changed even this feature is GAed.
2. When node admin options in with adding a `recheck period`, (in another word, this is for those admins who only allow its users using`Always` ImagePullPolicy now), I prefer that restarting kubelet needs recheck, which will be more secure for them. The admin can control the secure/weak level using the recheck period. This is also why I think no checkpoint is acceptable.

liggitt commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -224,3 +312,68 @@ func applyDefaultImageTag(image string) (string, error) {
+		}
+		digest := m.ensureSecretPulledImages[imageRef]
+		if digest != nil {
+			ensuredInfo := digest.Auths[hash]
+			if ensuredInfo != nil && ensuredInfo.ensured && ensuredInfo.lastEnsuredDate.Add(m.pullImageSecretRecheckPeriod.Duration).After(time.Now()) {
```
> If we have a boolean configuration for cross pod checking that can be set independently would we then default to some recheck period (cache invalidation) if not set or keep assuming image access for the entry in the cache till the kubelet is running (current in-memory impl)?

Not sure... I'd probably default both to current behavior and let admins opt into things that could break or disrupt currently working setups, but I hadn't thought about it for long.

mikebrow replied:
good point... let's add another config for the ensured by cross-pod checking option. I like the name.

What do you think about the default for the cross-pod checking option? Same as for the recheck period (defaults to off?)

liggitt replied:
> What do you think about the default for the cross-pod checking option? Same as for the recheck period (defaults to off?)

Yeah, my starting position is *always* to default to settings that won't make an admin that failed to read 40 pages of release notes come hunt us down because we broke them on upgrade

mrunalp replied:
sounds good. We will have to document that setting the booelan and not the recheck period could have implications :)

pacoxu replied:
If recheck is disabled(with recheck duration is 0s), a new pod with same hash should be checked again.

It sounds a valid corner case.

haircommander commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -136,12 +186,31 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
+
+	var pulledBySecret, ensuredBySecret bool
+	if present && utilfeature.DefaultFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) {
+		pulledBySecret, ensuredBySecret, err := m.isEnsuredBySecret(imageRef, spec, pullSecrets)
```
this looks like we're shadowing pulledBySecret and ensuredBySecret. Shouldn't this be `=`?

mikebrow replied:
nod looks like a regression after adding the earlier feature gate check...

pacoxu replied:
Fixed.

haircommander commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -164,14 +233,37 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
+				}
+				digest.Auths[imagePullResult.pullCredentialsHash] = &ensuredInfo{true, time.Now()}
+			}
+			defer m.lock.Unlock()
+		}()
```
why is there a different closure and why are we deferring at the end of it? I would expect the defer to go with the Lock() above, but really I think it'd be better without the closure and with the explicit lock/unlock directly inside the if scope

pacoxu replied:
My fault. Updated.

rphillips commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -224,3 +316,70 @@ func applyDefaultImageTag(image string) (string, error) {
+		return pulledBySecret, false, nil
+	}
+	// if the pullImageSecretRecheckPeriod is 0, we don't need to check the secret again
+	if m.pullImageSecretRecheckPeriod.Duration == 0 {
```
This should be `m.pullImageSecretRecheckPeriod.IsZero()`.

pacoxu commented:
in current design, to keep old behavior, we should use 'recheckperiod != 0 && feature gate enabled'

pacoxu committed:
if pullImageSecretRecheckPeriod>0 and KubeletEnsureSecretPulledImages is enabled, do recheck logic
- if recheck is enabeld, recheck no auth image(pullCredentialsHash=="") as well

haircommander commented on pkg/kubelet/images/image_manager.go:
```diff
@@ -164,14 +233,37 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
+				// successful pull no auth hash returned, auth was not required so we should reset the hashmap for this
+				// imageref since auth is no longer required for the local image cache, allowing use of the ImageRef
+				// by other pods if it remains cached and pull policy is PullIfNotPresent
+				delete(m.ensureSecretPulledImages, imageRef)
```
I kinda feel like pullCredentialsHash == "" should be handled in a separate mechansim. As in: if we remove images with an empty credentials hash from this map, then we treat the first pull after kubelet restart and first pull after recheck period as `pullCredentialHash == "'"`. Instead, can this be skipped or even add a "" entry in to the maps, and then we treat not present in map as always recheck?

pacoxu replied:
If the image needs no hash at first and then later needs auth, this would be a problem.

If recheck is enabled, the recheck for no auth image seems to be needed as well.

pacoxu replied:
https://github.com/kubernetes/kubernetes/compare/7a120ec8850900a419db335db37cf7063cba3b3f..51f8822fe18a3b044c76d866a765c8109b32b403 the latest update .

mikebrow replied:
@pacoxu 
What do you think about storing an "anonymous" entry instead of starting over.. when anonymous is successful.. If we go that route we can check for known anonymous success first then if not in the list check for each hash.. With duration checking enabled can also recheck anonymous..

mikebrow replied:
iow isEnsuredBySecret() needs to be improved to isEnsured() // By Secret || Anonymous  .. when anonymous worked for a particular registry that image is now in cache and thus any private pull with auth is moot, at least until another policy says recheck it now..  

I'm taking the presumption that we are not going to check if an image was pulled from registry a or registry b.. the image in the k8s.io image cache is still the same bytes either way..

pacoxu commented:
/hold
for there are still some things not clear.

/milestone clear

pacoxu commented:
/close 
ref https://github.com/kubernetes/kubernetes/pull/125817
